### ISSUE TITLE
typo in README.md - strutural

### ISSUE THREAD

**Issue by igortru** [steineggerlab/foldseek#155]
In default Foldseek uses its local 3Di+AA strutural alignment ...

related question:
is it possible separate this function into standalone call?
I have 2 AA sequences + 2  corresponding 3Di sequences , but I don't have pdb files
I just want to look on alignment..
Thank you in advance!

**Comment by martin-steinegger** [comment]
Sorry I do not fully understand the question. What do you mean by standalone calls?

**Comment by igortru** [comment]
Martin:
How exactly local 3Di+AA structural alignment implemented?
what is input , what is output? is it possible separate this procedure from foldseek databases structures?


it is all about optimization:
as I understand, whole foldseek server project was implemented as fast and cheap way search structural 
homologies in very large databases.

just continue this line:
alphfold2 is good butrequires huge amount of computational resources.
esmfold - very fast, foldseek use it on the fly , but have restriction 400aa
question : how extend it  to 2-3K AA? one way - use A100 with 80Gb, very expensive.

if I use V100 with 16GB , I have cutoff 700AA - and problem is MEMORY, not time.
I am ready exchange space on time (probably it is question for EsmFold developers)
it can be slower , but long proteins need to folded.
as minimum, output can be  just full length 3Di sequence

then this sequence can be searched against existing foldseek databases.

as experiment I run 700AA fragments with step 1AA , retrieve 3Di sequences, align them, 
calculate consenus,
 Then I want  align it (using foldseek algorithm) against existing 3D-structures  (in 3Di/AA space)
for now I have only 3Di scoring matrix.
 
realted problem:
what is the best way split long sequences on short fragments and analyze it using small amount GPU memory?
is it enough just trivial 400AA fragments with 200AA step? or something more intellectual can be implemented.

**Comment by milot-mirdita** [comment]
I am not sure how structure prediction and structure search relate to in your question. Could you please clarify?

If you want a faster structure predictor than the normal AlphaFold2, you can use [ColabFold](https://github.com/sokrypton/ColabFold). ColabFold implements many optimizations to speed up AlphaFold2 and also reduce memory requirements somewhat.

Foldseek does not use an GPU resources.

**Comment by igortru** [comment]
foldseek web server use esmfold for structure prediction when as input provided fasta.
it has restriction 400aa.

what is the best way search long sequences 1-2-3 K
using foldseek server ?
cut sequences on 400aa fragments and then somehow merge  search results on client side?

Is it possible make such merges inside foldseek on the fly and remove restriction on sequence length?

it is not about high quality 3d-structure prediction, but about searches  inside foldseek large databases using “local” (400aa) 3d-structures.

**Comment by milot-mirdita** [comment]
We are working on a different solution for sequence-to-structure searches currently. The limitations of the ESM API will probably remain as their are. I don't have a better solution for you currently, than predicting the full long structure with ColabFold first.

**Comment by igortru** [comment]
https://aibusiness.com/nlp/meta-lays-off-team-behind-its-protein-folding-model

**Comment by igortru** [comment]
I would say,  my interest  is not sequence to structure search , but  fast and cheap method of sequence to sequence “2.5D” search using as intermediate 3Di alphabet.

I have two  sequences which presented in AA and 3Di alphabet,no foldseek database, just sequences.

How I can align them using foldseek?
is it possible separate  aligner as standalone tool?
is it possible convert my sequences into  foldseek database?

Right now I am developing tool (ML)  which will allow direct conversion from AA to 3Di.
I don’t need very high quality conversion procedure… It is just proof of concept :
I think, such distilation from AA-alphafold-3Di is really possible.

**Comment by martin-steinegger** [comment]
Please have a look at ProstT5, which is a transformer that can do this conversation. It’s a translator between AA and 3Di. https://www.biorxiv.org/content/10.1101/2023.07.23.550085v1

**Comment by igortru** [comment]
Thank you! I’ll try…
last step:  how align  such sequences  in two spaces?

**Comment by igortru** [comment]
ProstT5 is working as expected,
not perfect , but acceptable.
now only question  , how create foldseek database
using only AA and 3Di sequences  without pdb  files.
I am going to process using ProstT5 hundreds thousands proteins and then cluster them using foldseek.
As I understand _ca file could be empty.

**Comment by martin-steinegger** [comment]
You would need to generate three tsv files 
(1) AA
```
1 MAVS
2 MAAAV
```
(2) 3Di
```
1 VVPL
2 VLAT
```
(3) header 
```
1 protein1
2 protein2
```

Then you need to convert them into three dbs using our `tsv2db` command. The dbs should be named `dbname`, `dbname_ss` and `dbname_h` and should have the db type `aa`, `aa` and `generic` respectively.

**Comment by igortru** [comment]
it looks like , ProstT5 from  the same class of models as EsmFold, but has higher requirements to GPU memory, my V100 16GB just not enough.

on first glance it is equal  to AA-EsmFold-Foldseek.
using EsmFold I can easily process 17K sequences with sequences length less than 700AA,
Using ProstT5  I can process only two short sequences at once.
three sequences throw exception -not enough memory (half model)

Thank you for Tsv2db , it is exactly what I need.

**Comment by igortru** [comment]
Martin,
it looks like ProstT5 has problems with some sequences.

compare ProstT5 vs EsmFold+Foldseek
genbank accession WGH50207.1 
WGH50207.1 MIDITGTPENLTSQSLYIELYKTTKIHITSTMNQNGVLETDENNNNLMSKLNVGDVKIDSILGVTSDYLLLYTSDKNKYMLINSTSSNTSYKMVWNKQLMYKIGAKTLYNQTLHSDYTVDIPANYSLFALYINNNRVKDSEYSISGQTITLSQKLRNKLDLVYNTVNMVVYKTDQTITTSIELHYLSVKEFFNSITFRDESYFTTNNIREHINYESLKYDEEITNETFGNERRKEIIRTNMIPKLTLSYFVGENFTDFGNELLSKTFRVIVNHPNTKTMEIYPNCIVTSGYKKDYAKEKNTVEVEIQCDGRYEIQFGTQPQLINRPYNSGTYNNGIYG ddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddpddddddddddddddddddddpdddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddd

WGH50207.1 MIDITGTPENLTSQSLYIELYKTTKIHITSTMNQNGVLETDENNNNLMSKLNVGDVKIDSILGVTSDYLLLYTSDKNKYMLINSTSSNTSYKMVWNKQLMYKIGAKTLYNQTLHSDYTVDIPANYSLFALYINNNRVKDSEYSISGQTITLSQKLRNKLDLVYNTVNMVVYKTDQTITTSIELHYLSVKEFFNSITFRDESYFTTNNIREHINYESLKYDEEITNETFGNERRKEIIRTNMIPKLTLSYFVGENFTDFGNELLSKTFRVIVNHPNTKTMEIYPNCIVTSGYKKDYAKEKNTVEVEIQCDGRYEIQFGTQPQLINRPYNSGTYNNGIYG dpppppdpppclpkdkdwdfpdadaddapddqdpvnfdqagpvrhgclvvddqppfyqdphppppdtwtfhddpppsgtdtadcddppdqrhpppdpddpqpqarfdkdkdaqdpqrkdfadpqkdfpwkdkpndtddpvqwdddprmihghpvvvvvddnvdiimitgihgnvppspspppppdpppqnsvvvpppvdscppvpdpqsgdwdkdwfdfpqpqpqdddppdpdrdrdgddrwtkmktkgfddppddppvvvqqpgfikmwmadqvqqkikiagrkhfdddwdwdpdvpttmtmtmiigpddidmdhddgppppppppdpppppppppd

**Comment by mheinzinger** [comment]
Hi @igortru , just quickly jumping in here as this rather seems to be an issue with ProstT5.
I also observed that ProstT5 generates sometimes very repetitive sequences. Similar to NLP, this is most likely due to non-optimal parameters for the [generation config](https://huggingface.co/docs/transformers/generation_strategies).

An intermediate solution from our end is to avoid the decoder-based generation of 3Di sequences altogether and instead simply put a CNN on top of the Encoder of ProstT5. We have now a [new a script/prediction_head](https://github.com/mheinzinger/ProstT5/blob/main/scripts/predict_3Di_encoderOnly.py) which allows you to do exactly this (go from input:FASTA(AA) to output:FASTA(3Di) using only a CNN on top of the encoder of ProstT5.
This solution bypasses repetitive predictions by replacing the sequential token-by-token generation of the decoder by a single forward pass through the encoder. This should also be much faster :)
Just in case, I gave the sequence you posted above a shot and the prediction based on the CNN/encoder solution looks already much less repetitive (still some D's, though):
`>WGH50207_1
DDDDAADDDDQDFKKKKKFFFDDDKKAFDWDADDVQDFDDDPVRPVVCVVDDDDPDDFDDFPDDDQFWTWTAGPVRQIIDTDDDPDDDDMIIIMGDDDIKDDDDKDDDDWDDDDQKDKDDDDPQKDWKWKDKLNHTDDPVQWDDDGRMIGGDPVVSVVDDPVGIMIIMWIGGDPDDDMIMMMIGMMTDDDAFWWKDWDDDDIDDDPDDDDDKDDDDKDWDKDWDDDDPDDDIDIDIDTDDTKIKMKTKGWDDPDDVVVVVVQAPDWTWMWTQDPVQRKIKIFGGWHWPDDDDDDDDPDTRMIMTMITGPDDIDDDDDDDPDGDPDDPPPDDDDPDPPD`

**Comment by igortru** [comment]
Thank you , Michael!!
The result looks very interesting...
I have run small experiment
from   ProstT5 3D-foldings training set I have retrieved structures with
length 300AA , it is about 21K proteins.
and "fold" them using predict_3Di_encoderOnly.py - 2
then I have retrieved 3Di sequences from the training set using foldseek.
-1

 each pair of 3Di sequences I have aligned  using needall with mat3di.out
(foldseek 3di scoring matrix) -gapopen 11.0 -gapextend 1.0(probably can be
adjusted)
in attachment you can find self scores distributions (1,2 - they are not
very different), and score distribution when I align 1 vs 2.
I see aa lot of space for improvements.

If you provide training code for your simple CNN, I will be very thankful.
I want to check my CNN, how it works on top of ProstT5.
 I can try to do it myself , but I am not very proficient in pytorch.

Thank you in advance!
Best regards!

iTolstoy

**Comment by mheinzinger** [comment]
Sure, we have just used [bio_trainer](https://github.com/sacdallago/biotrainer/) together with the [autoeval version of the NetSurfP-2.0 secondary structure dataset](https://github.com/J-SNACKKB/autoeval) for training. We used ColabFold 3D structure predictions to derive 3Di to avoid missing residues. 

on your analysis: Hm, I have to admit that I would need some more details on what you did exactly. From what I understand, you take a random subset of 21k proteins with corresponding 3Di/AA data from our ProstT5 training data (with random being only proteins with 300AAs) and then you align those against themself (which you label 1) and against the 3Di prediction using the CNN (which you label 2). Those look similar when being aligned against themselves but different when being aligned against each other. Is this correct? - which metric do you report exactly? - Just stupid Q: you are sure that those sets are identical? - Is the sorting of hits still correct, i.e., is a self-hit between the two sets ranked higher than any other pair? - Could you provide E-value distribution using FoldSeek? (I think it would be best to really use Foldseek for this sort of alignment to avoid any unintended side-effects)

**Comment by igortru** [comment]
Hello,Michael

below raw data 
/https://ftp.ncbi.nlm.nih.gov/genomes/Viruses/TEST.3di
as identifier I am using alphafold accessions
_1 : 3Di sequence from Alphafold prediction using foldseek
_2: 3DI sequence which predicted your script using as input protein fasta
all 21K pairs in one file

https://ftp.ncbi.nlm.nih.gov/genomes/Viruses/TEST.score
needleall output:
pair of accessions , alignment length, alignment scores
you can see  four pairs for each accession :  _1 _1, _1 _2 , _2 _1 , _2 _2

How exactly you calculate loss for your CNN?
one possible way  I just  provided above

**Comment by mheinzinger** [comment]
We use a simple cross-entropy loss on each residue. So no alignment etc but just simple token/residue-classification

**Comment by igortru** [comment]
how residues  represented ?
as one hot or just number?
I really interested to look on your exact code
I have problems with my model training when  I average loss along 3di sequences

**Comment by igortru** [comment]
Michael, I can't find where this training dataset located:

[autoeval version of the NetSurfP-2.0 secondary structure dataset](https://github.com/J-SNACKKB/autoeval)

**Comment by igortru** [comment]
I am a little bit confused  about biotrainer source code
when I open  config.yml  directly on github site 

https://github.com/sacdallago/biotrainer/blob/60aaa0e70c28200c6a378e26af6dd47ee37a66d6/examples/residue_to_class/config.yml

I see
 embedder_name: prottrans_t5_xl_u50
 
 but when I 
 git clone https://github.com/sacdallago/biotrainer.git

 I see different config.yml
 
 embedder_name: one_hot_encoding
 
 Michael, could you explain difference?
